Proposal: query path tenancy (#6320)
thanos-io/thanos: docs/proposals-accepted/202304-query-path-tenancy.md

---
type: proposal
title: Tenancy awareness in query path
status: accepted
owner: douglascamata
menu: proposals-accepted
---

## Tenancy awareness in query path

* **Owners:**
    * [@douglascamata](https://github.com/douglascamata)

* **Related Tickets:**
    * [Querier: Added native multi-tenancy support](https://github.com/thanos-io/thanos/pull/4141).
    * [Proposal: Native Multi-tenancy Support](https://github.com/thanos-io/thanos/pull/4055)

* **Other docs:**
    * [Multi-tenancy in Thanos](https://thanos.io/tip/operating/multi-tenancy.md/)

> <!--

Reviewer: Can we remove these comments? Causes no diff in rendering, but makes it difficult to review. 🙂

### Pitfalls of the current solution

<!--
What specific problems are we hitting with the current solution? Why it’s not enough?

*For example, We were missing a consistent design doc template, so each team/person was creating their own. Because of inconsistencies, those documents were harder to understand, and it was easy to miss important sections. This was causing certain engineering time to be wasted.*
-->

The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools or proxies. For example, it's impossible to determine which tenants are experiencing high latency or high error rates.

## Goals

<!--
Goals and use cases for the solution as proposed in [How](#how):

* Allow easy collaboration and decision making on design ideas.
* Have a consistent design style that is readable and understandable.
* Have a design style that is concise and covers all the essential information.
-->

> * Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path. These features include, but aren't limited to, the following:

Reviewer: ```suggestion
* Allow the query path components to be configurable to identify tenants, opening the way to the implementation of per-tenant features on the query path. These features include, but aren't limited to, the following:
```

  * Per-tenant observability.
  * Per-tenant settings. For example, having different limits per tenant, which is a common request. 
  * Enforce presence of one or more tenant labels in queries.

### Audience

<!--
If this is not clear already, provide the target audience for this change.
-->

> Any team running Thanos in a multi-tenant environment.

Reviewer: ```suggestion
Any team running Thanos in a multi-tenant environment i.e, Monitoring-as-a-Service use cases.
```

## Non-Goals

<!--
* Move old designs to the new format.
* Not doing X,Y,Z.
-->

> * Add multi-tenancy to Thanos Ruler. It's not needed. One will still be able to point their Ruler to the multi-tenant Query Frontend, effectively making it multi-tenant. 

Reviewer: It is needed, as rule configuration needs to be treated separately. You can select a series of a particular tenant, but you must also add back that label to the newly generated series, which makes it a bit more complex. Also, in stateless mode, Ruler needs per-tenant config to write to tenant TSDBs. 

Maybe let's skip mentioning need here, and just mention that it is out of scope for this proposal? 🙂 
```suggestion
* Add multi-tenancy to Thanos Ruler. 
```

Author: Rules can be already configured to add the tenant label, no @saswatamcode? For example, in our deployments we are putting the results of recording rules under specific tenants, otherwise it would be a complete mess. The write path is already prepared for this.

The proposal applies only the the query path and in reality there's a problem there. If the Ruler doesn't indicate a tenant through the HTTP header of a query, the default tenant will be picked, and then the tenant label enforcement and verification will fail because of the mismatch. **So in fact we need to update the ruler**.

Reviewer 2: > So in fact we need to update the ruler.

Can we clarify off the back of this conversation what needs to be updated then?

Author: Alright, I did some investigation. We import the rule evaluation code directly from Prometheus, which owns the format of the rule configuration files and all the logic behind it. This makes fixing the Ruler a harder problem. We have two paths forward:

1. On a Querier with multi-tenancy enabled (`--querier.tenant-header` present), change the behavior when the header isn't sent. So instead of using the default tenant it would skip tenancy check. It seems like a bad idea to have a different behavior from Receive in the same situation (Receive considers requests without the tenant header to be meant for the default tenant). **This is not what I think we should do**.

2. Leave the Ruler out of the proposal. Meaning it will need to have its own pool of Queriers with multi-tenancy disabled to keep working normally. A new proposal can be created in the future to outline how we handle multi-tenancy in Ruler.

So, for now, I prefer to go with 2. 

WDYT, @saswatamcode?

Author: FYI I'll update the diagram to reflect my preference for option 2.

Reviewer: Let's go with option 2 for now! :)

Author: > Can we clarify off the back of this conversation what needs to be updated then?

@PhilipGough Ruler would have to know to which tenant a given rule belongs to, so that when sending queries it includes that tenant's name in the tenant header. This means we would need to tap into the rule configuration format and evaluation engine, as both are  unaware of tenancy, as they belongs to Prometheus.

Author: btw, continuing on ⬆️ , Ruler would also need to know the tenant for the local (TSDB, requires no change) or remote (remote write, requires change to send tenant header) storage.

* Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.

## How

<!--
Explain the full overview of the proposed solution. Some guidelines:

* Make it concise and **simple**; put diagrams; be concrete, avoid using “really”, “amazing” and “great” (:
* How you will test and verify?
* How you will migrate users, without downtime. How we solve incompatibilities?
* What open questions are left? (“Known unknowns”)
-->

> * Implement a mechanism to allow incoming requests to specify the tenant being queried using an HTTP header. This applies to all components in the query path. We follow the example of Thanos Receive, which uses the `--receive.tenant-header="THANOS-TENANT"` flag to configure the tenant header, adapting its name to each component.

Reviewer: Can we specify the flag that is planned to be added to Querier?

Author: This is described by the text: we take the same flag as Receive's, `--receive.tenant-header="THANOS-TENANT"`, and adapt its name to the component. 

So in the Querier, the flag name will be `--querier.tenant-header`, in the Store Gateway it'll be `--store.tenant-header`, and so on.

Reviewer 2: This seems fine, perhaps worth just adding to the proposal Douglas to ensure clarity.

* The tenant header value from a given request should travel downstream to all the components being called, so that it can be added to their metrics, traces, and logs without requiring duplicated/extra work to re-parse the query. This also applies to gRPC calls.

> * Implement a command line argument in the Querier component that will indicate with label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name. The default behavior, when the flag as an empty string as value, is to not identify tenants. 

Reviewer: ```suggestion
* Implement a command line argument in the Querier component that will indicate the label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name. The default behavior, when the flag is an empty string, is to not identify tenants. 
```

  * The label verification and enforcement should be done by reusing prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141). There's no reason to (re)implement something specific and special for Thanos.

>   * This behavior should be implemented as part of the base logic of both the HTTP and gRPC query APIs, before the query is handed to the query engine. This allows users managing complex Querier trees can choose where they want this logic to be enabled. 

Reviewer: What about StoreAPI, which is mentioned in the architecture diagram? Also, maybe worth mentioning how the header propagation flow would work for different architectures? 🙂

Author: There's no need to add or change anything at all in the Store API definition. As mentioned there in the text, the HTTP header will be passed even on gRPC requests, which in the end are also HTTP requests. So it'll handle the header just as any other server will: check if it's present and use it as an extra label in its own exported metrics.

Author: The behavior of the Store API server doesn't change at all from the point of view of a Store API client. The only component that will check the tenant header value against the tenant label selector is the Querier.

* Update metrics exported by the components in the query path to include the tenant label when it's available. 

> * Implement a tenant selector in the Query Frontend UI, which should communicate the tenant to Query Frontend using the HTTP header.

Reviewer: I'm not entirely sure I understand this. Why have a selector in the UI? 
Likely tenants will be authenticated in some layer above Thanos, and that is where tenant HTTP headers would be added (even for cross-tenant cases which are out of scope here).

The UI would just be proxied, and would make requests with the header present.

Author: > Likely tenants will be authenticated in some layer above Thanos

This is completely optional, we cannot assume everyone will have authentication. One might still want to run queries with multi-tenancy checks enabled even without an authentication layer, most importantly because it enabled observability of queries based on tenant all the way throughout the query's path.

Reviewer 2: Thinking out loud and taking the scenario where authn/authz happens before the user reaches the UI, does adding a selector here pose any risk to leaking anything?

Author: @PhilipGough it does. But this demands some higher level thinking: is it Thanos' responsibility that there is no data leak? IMHO, it shouldn't be.

If you get to the point of seeing a Thanos Query (Frontend or not) UI, you can easily watch you network tab and craft a request that will return data from other tenants. We can only avoid this by making Thanos authn aware, which is not something Thanos wants to do.

Author: I would change the wording though. In the query UI it wouldn't be a selector, but instead a free-type textbox.

Reviewer 2: I agree. I just wanted to call out the issue

<figure>

> <img height="50%" src="../img/query-path-tenancy-proposal-diagram.svg" width="50%"/>

Reviewer: Probably you can increase the size of the diagram a bit more, so that it is legible on the website (see preview).

1. This is why not solution Z...
-->

### Alternative implementations

1. Apply verification and enforcement logic in the Query Frontend instead of Querier.

The Query Frontend is an optional component on any Thanos deployment, while the Querier is always present. Plus, there might be deployments with multiple Querier layers where one or more might need to apply tenant verification and enforcement. On top of this, doing it in the Querier supports future work on using the [new Thanos PromQL engine](https://github.com/thanos-community/promql-engine), which can potentially make the Query Frontend unnecessary.

### Alternative solutions

1. Use a proxy to enforce and identify tenants in queries.

While this could work for some of the features, like exporting per-tenant metrics, it would have to be inserted in front of many different components. Meanwhile, it doesn't solve the requirement for per-tenant configuration.

2. Use a separate query path for each tenant.

This incurs in a lot of wasted resources and demands manual work, unless a central tenant configuration is used and a controller is built around it to automatically manage the query paths.

> ## Action Plan

Reviewer: Maybe we can add steps for implementation in this section?

Author: This section is weird, because in the "How" section the steps are already described. I don't see value is describing steps in much more (technical) details here.

Reviewer 3: I guess here you can link tickets or make a list of items if you plan to break down the work somehow.

---
type: proposal
title: Tenancy awareness in query path
status: accepted
owner: douglascamata
menu: proposals-accepted
---

## Tenancy awareness in query path

* **Owners:**
  * [@douglascamata](https://github.com/douglascamata)

* **Related Tickets:**
  * [Querier: Added native multi-tenancy support](https://github.com/thanos-io/thanos/pull/4141).
  * [Proposal: Native Multi-tenancy Support](https://github.com/thanos-io/thanos/pull/4055)

* **Other docs:**
  * [Multi-tenancy in Thanos](https://thanos.io/tip/operating/multi-tenancy.md/)

This design doc proposes to add tenancy awareness in the query path.

## Why

> In a multi-tenant environment, it's important to be able to identify which tenants are experiencing issues and configure (e.g. with different limits) each one of them individually and according to their usage of the platform so that the quality of service can be guaranteed to all the tenants.

Reviewer 3: Nit: the long form is usually preferred for written communication. 

```suggestion
In a multi-tenant environment, it is important to be able to identify which tenants are experiencing issues and configure (e.g. with different limits) each one of them individually and according to their usage of the platform so that the quality of service can be guaranteed to all the tenants.
```

### Pitfalls of the current solution

> The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools or proxies. For example, it's impossible to determine which tenants are experiencing high latency or high error rates.

Reviewer 3: This is not fully correct since we can run different queriers for individual tenants. It is worth rephrasing the sentence a bit to indicate that such a setup is more complex and resource inefficient because we cannot pool resources.

Author: Sure thing. I'll tackle this together with the other comment you made on the "Goals" section, @fpetkovski, as they are closely related. 

Indeed one can achieve the goals with at least one querier per tenant, but I think only partially. For example, the Query Frontend and Store Gateway metrics will still lack per-tenant visibility, unless you also additional deploy at least one instance of Query Frontend and Store Gateway per tenant.

Author: In the end, to have per-tenant visibility and configuration on the query path **today**, you need to deploy a full query path per tenant. It shows how strong the lack of "true" multi-tenancy is.

Author: @fpetkovski I tried to incorporate your feedback into the writing. Please, give it a quick read when you can and let me know if it's better.

## Goals

* Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path. These features include, but aren't limited to, the following:
  * Per-tenant observability.
  * Per-tenant settings. For example, having different limits per tenant, which is a common request.

>   * Enforce presence of one or more tenant labels in queries.

Reviewer 3: We can achieve these goals by having separate query instances for individual tenants. A valid goal would be simplifying management for multi-tenant setups and pooling resources to reduce waste.

Author: Hey @friedrichg, can you give us some good pointers for this implementation on the Cortex side? I'm not familiar with Cortex and its features, so I'm very unaware of how it implements things.

If it aligns well with how we want to move forward with this, it'll be great to reuse some code (be it importing or copying) so that both projects benefit from it. I know there are some design differences between the two projects, so let's see what can be done and reused.

Reviewer 4: Yea, I think it's good point to mention Cortex way of dealing this as inspiration, but there is nothing wrong in revisiting this feature and checking if there is anything we could improve before implementing.

Ideally, we could have the same solution, so the tools that works on top of Cortex will work on top of Thanos. That's fair advantage to adopting similar patterns. Another advantage is easier migrations between projects, helpful if we will try to unify projects at some point.

Can we mention this somewhere in the doc (advantage and Cortex pattern)? Thanks!

* Implement a mechanism to allow incoming requests to specify the tenant being queried using an HTTP header. This applies to all components in the query path. We follow the example of Thanos Receive, which uses the `--receive.tenant-header="THANOS-TENANT"` flag to configure the tenant header, adapting its name to each component. So in the Querier, the flag name will be `--querier.tenant-header`, in the Store Gateway it'll be `--store.tenant-header`, and so on.
* The tenant header value from a given request should travel downstream to all the components being called, so that it can be added to their metrics, traces, and logs without requiring duplicated/extra work to re-parse the query. This also applies to gRPC calls.
* Implement a command line argument in the Querier component that will indicate with label name that should be used to enforce tenancy. Again, we follow the example of Thanos Receive, which uses the `--receive.tenant-label-name="tenant_id"` flag to achieve this, and adapt its name just as described on the first point (changing the `receive` part to `querier`, `store`, etc). The default behavior, when the flag as an empty string as value, is to not identify tenants.

>   * The label verification and enforcement should be done by reusing prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141). There's no reason to (re)implement something specific and special for Thanos.

Reviewer 3: I don't think we should be modifying incoming promql expressions since that can have unexpected side effects, especially with functions like `absent`. 

I would suggest enforcing the tenant in the storage querier by directly adding a separate label matcher: https://github.com/thanos-io/thanos/blob/main/pkg/query/querier.go.

Reviewer 3: Ah just realized that enforce matchers does exactly that. Using #4141 as a starting point makes sense 👍

### Pitfalls of the current solution

The current lack of tenancy awareness in Thanos' query path makes it impossible to investigate issues related to multi-tenancy without the use of external tools, proxies, or deploying a full dedicated query path for each tenant (including one Query Frontend if the setup needs it, one Querier, and one Storage Gateway, to be as complete as possible). For example, it's impossible to determine which tenants are experiencing high latency or high error rates

## Goals

* Allow the query path components to be configurable to identify tenants, opening the way to the implementation per-tenant features on the query path without needing to run multiple copies of the components, each dedicated to a single tenant. These features include, but aren't limited to, the following:
  * Per-tenant observability.
  * Per-tenant settings. For example, having different limits per tenant, which is a common request.
  * Enforce presence of one or more tenant labels in queries.

### Audience

> Any team running Thanos in a multi-tenant environment. For example, a team running a monitoring-as-a-service platfor. For example, a team running a monitoring-as-a-service platform.

Reviewer 5: ```suggestion
Any team running Thanos in a multi-tenant environment. For example, a team running a monitoring-as-a-service platform. For example, a team running a monitoring-as-a-service platform.
```

## Non-Goals

* Add multi-tenancy to Thanos Ruler. 

> * Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.

Author: @GiedriusS: it will still be possible, by default, to keep backwards compatibility. When the tenancy-enforcement feature is turned on a Querier, then it won't be possible for requests arriving there. But you can have another set of Queriers with the feature disabled and multi-tenant queries will work, just like today. o/

> * Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.

Author: Cross-tenancy is definitely a future goal of the project and a good "follow up", but I didn't plan this proposal to go deep into its details. I don't think it'll block in any way a cross-tenant query implementation. Cross-tenant queries will require changes to the logic behind the label enforcement part, but there's no architectural decision that will need to be undone when the time comes. 

Totally we will need to develop more on these two aspects of tenancy you mention, @bwplotka. Plus we will need some configuration that will work kinda like an authorization for cross-tenant query (i.e. tenant A can query B and C; tenant B can only query C, etc).

> * Implement cross-tenant querying. It poses the question about how to track metrics of a multi-tenant query, probably requiring new logic for query splitting and/or separation of the concepts of query initiator tenant and query target tenants. To exemplify: if tenant A executes a query touching their own data and tenant B data (imagine a label matcher `tenant_id=~"A|B"`), how does Thanos export a request duration metric about this query? This is out of scope and an specific proposal can be created for it.

Author: @GiedriusS: I am trying to ensure we keep backwards compatibility and the possibility of running queries with and without tenant-enforcement in a cluster. Maybe we can collaborate later on a proposal for multi-tenant query with the tenancy enforcement on? It's not a goal of this proposal to outline how this would work. What do you think? :D

### Alternative solutions

1. Use a proxy to enforce and identify tenants in queries.

While this could work for some of the features, like exporting per-tenant metrics, it would have to be inserted in front of many different components. Meanwhile, it doesn't solve the requirement for per-tenant configuration.

2. Use a separate query path for each tenant.

This incurs in a lot of wasted resources and demands manual work, unless a central tenant configuration is used and a controller is built around it to automatically manage the query paths.

## Action Plan

1. Query Frontend

>    1. Implement the `--query-frontend.tenant-header` (default is `""`") and `--query-frontend.default-tenant` (default is `"default-tenant"`) flags. Forward the tenant header to all the configured downstream query endpoints, if it is present.

Reviewer 3: Configuring the tenant in multiple places can be a big source of confusion for users. Can query frontend simply propagate all headers? I would prefer it we enabled tenancy on the query path using one flag in one component.

Author: We need it to be configurable in multiple places because this header has to go all the way down to the Thanos Store, not only from Query Frontend to Query. It also has to be transmitted from Querier to Store Gateway in setups without the Query Frontend and from Querier to Querier in setups with multiple Querier trees. 

We need every component in the query path to be able to identify the header containing the tenant name, so there's no escape configuring this everywhere. If we configure it only in the Query Frontend, how does the Querier know which header includes the tenant's name?

I'm against doing a blanket forwarding of all headers. It seems like a potential source of security issues or unknown unintended consequences. WDYT?

Author: Basically, every component has to agree on which header is being used to identify the tenant and we have nothing to facilitate centralized configuration at the moment. 

Our only option is to hardcode the tenant header name to be always the same and impose it on all the users of Thanos that want to use this feature.

Author: Ah, there's a good reason for not hardcoding the tenant header's name: we don't do this in Receive. It seems better to me that we keep the same behavior in the query and write paths. Our default tenant header name is `THANOS-TENANT`.

Author: @fpetkovski it does work, but as I mentioned before: is not enough, not necessary, and breaks the pattern of the header that is used in Receive. 

It's not enough because: 

- Query Frontend has no awareness of the Store API at all (only the Queriers do, as they interact with Store Gateways), so how would I be able to observe it per-tenant and have per-tenant features there? It has to receive the tenant indication through the header or modifying the Query API. If we modify the Query API here, we won't be compatible with people simply pointing their Grafana datasources to a Query Frontend -- how would they add this extra "tenant" parameter to their requests? 
- Queriers can also receive requests via the Query API. So we also have to modify the Query API here or accept the header at this level too. 

It's not necessary because:

- A header can do the same thing, without any change to the Store API. Cortex and Mimir are proof that this works well. 
- A header works fine with Grafana datasource configurations.

The last point is self-explanatory: in our ingest path the tenancy is indicated through a (configurable) header on Receive. Why do it differently in the query path? 

So in the end, why would I prefer to modify the Query and Store APIs if using and propagating a single header is simpler while achieving the same features and covering every components with a single standard?

Reviewer 3: Sorry but I feel this discussion is unfocused and unproductive. The original suggestion is to find a way to reduce the number of parameters that have to be configured to enable tenancy and I would appreciate it if we tried to work towards a solution together. My last comments were about the new `--store.tenant-header` which would be added to the store gateway because you preferred to keep the discussion in one thread.

Reviewer: I agree with @fpetkovski here too. I don't think configuring parameters on every single Thanos component is the right way to go about this, and adding a field to StoreAPI would be way more simple. Let's work towards a solution on those lines. 

We already treat Ruler differently and out of scope for this proposal, we can do the same for Query frontend for the time being. Let's first ensure tenancy on the base Querier -> StoreAPI implementation flow, and then expand to other special components.

Author: > The original suggestion is to find a way to reduce the number of parameters that have to be configured to enable tenancy

@fpetkovski I understand your intention, but my point is that in doing so we can't achieve everything included in the proposal. We already have lots of configurations in every component, with some of them having to be repeated in more than one place, like replica label configuration, for example. Why pick on this now? Is one CLI flag in 3 components worth the complications in implementation? Can't we leave configuration improvement for a follow up proposal on centralized configuration, for example? This way we unblock the implementation of the features, that's the core intention of the proposal, and can work on improving not only this but a lot of configuration spread over the whole project.

To move forward with your (including @saswatamcode's) requested changes to the proposal, we would already have to exclude the Query Frontend and go directly to a solution that involves modifying Query API  and Store API. That's considerably more risk, because of it's harder to remove or change something in the Store API (it strives to be as stable as possible) and our Query API will start to diverge from Prometheus'. 

With your proposal in mind, I would ask to please give some thoughts and elaborate on the following questions that weren't answered yet. For us to work together towards something and have a productive discussions, I need to understand what would be the solutions in your head for these things:

* How will Querier forward tenant identification to subtrees of Queriers? The queries are sent using Prometheus Query API via gRPC. 
* How will Querier identify tenants in setups without the Query Frontend? Here the communication is done via Prometheus Query API.
* How will Querier identify tenants in setups with the Query Frontend? Do we really want to forward **all headers** from Query Frontend to Query? What about potential security impact and unintended consequences I mentioned before? If we decide it's fine to forward all headers, how will I have settings and observability per-tenant in Query Frontend?
* How could we update the Query Frontend later on to do the same?

Author: Needing 1, 3, or any arbitrary number of CLI flags to achieve a good implementation we are happy with as is a consequence and not the objective. I believe the focus should be on achieving a solution as simple, comprehensive, and standardized across components, as possible. I understand that the configuration of Thanos components is getting complex and repetitive. That's a problem on its own and it deserves some improvements, rather sooner than later I would say.

Author: @bwplotka me and @fpetkovski had a call this morning to better discuss. So it’s alright, no hard feelings between us. In fact we came to even better points than what we talked about here. 😄

Either me or him will post the summary of the conversation later. We found a good middle point regarding both approaches to ensure compatibility with other tools (like Cortex and others), to have consistent behavior across components (how Receive behaves regarding its tenant header), minimize amount of config flags, and avoid modifying APIs.

Reviewer 3: We synced up yesterday via video call and agreed that tenant propagation can happen via a pre-defined header or field in the `SeriesRequest` message, which should remove the need for a flags in store components. 

There is still no good way to have per-tenant features in query frontend, so we might need to have a flag in both the querier and QFE if we want tenancy at the frontend level.

Reviewer 3: Actually now that I think about this, label injection would happen in the Querier, so only the Querier needs a flag to indicate the tenant label. If we go with @bwplotka's proposal to use pre-defined header names, Query Frontend just needs to forward the header to the Querier, so we won't need additional configuration options.

Author: @fpetkovski as we spoke yesterday, I think we **need** customizable header name because Receive's already offering that (through `--receive.tenant-header="THANOS-TENANT"`) and using two different header names depending on whether it's a query or a remote write creates confusion through inconsistency. 

I think it's valuable to keep the consistency here so that people that already know how Receive works will easily understand how the tenancy enforcement on the query path works.

The default header name will also match Receive's. So we will end up with:

  * `--querier.tenant-header="THANOS-TENANT"`. 
  * `--querier.default-tenant="default-tenant"`.
  * `--querier.tenancy="false"`.

When using the default header name, you will have to only add `--querier.tenancy="true"` to toggle everything on. The simplest case is gracefully simple to handle.

And if you are using a custom header name in Receive, you may (or may not) make the query path consistent with it by customizing `--querier.tenant-header`. The same applies to the default tenant setting.

cc @bwplotka

Author: As also mentioned by @bwplotka, the possibility of customizing the header name allows us to be more compatible with other tools, like Cortex, which facilitates interoperation and migration between tools.

Reviewer 3: That makes sense to me.

Author: @fpetkovski actually, we need that header name flag in the Query Frontend too. It cannot simply forward all headers. It needs to know what is the tenant header to include the information in its own metrics and for the per-tenant settings, as previously mentioned in our conversations.

   2. **In the UI**, add a textbox where one can type a tenant name. If the textbox has any content, that will be sent using in the configured tenant header name.
   3. Update the metrics exported by the Query Frontend to include a tenant label with the tenant indicated by the header.
2. Querier
   1. Implement the `--querier.tenant-header` (default is `""`") and `--querier.default-tenant` (default is `"default-tenant"`) flags. Forward the tenant header to all the downstream API calls (gRPC or HTTP, also using headers), if it is present.
   2. Implement the `--querier.tenant-label-name` flag. If it is not empty, verify and enforce the tenant label in the query before it is handed to the query engine. As previously mentioned by the proposal, use prom-label-proxy's [Enforce.EnforceMatchers](https://github.com/prometheus-community/prom-label-proxy/blob/main/injectproxy/enforce.go#L141).
   3. Update the metrics exported by the Querier to include a tenant label with the tenant indicated by the header.
3. Store Gateway

>    1. Implement the `--store.tenant-header` (default is `""`") and `--query-frontend.default-tenant` (default is `"default-tenant"`) flags. Use the tenant header to identify the tenant in the metrics exported by the Store Gateway.

Reviewer 3: Can't we use the tenant from request headers?

Author: How do we know the header in the request that includes the tenant's name if the header used to transmit this information can have its name changed? 

i.e. Query Frontend configured to use `X-User` as tenant header. How will Store Gateway know that `X-User` is the header being used for tenant identification and not anything else?

Reviewer 3: The querier knows which header is used for tenancy, and it can propagate this header in either the request or in a special gRPC/http header.

Author: This overlaps with our conversation in the other comment, let's solve that one first and then we come back to this one if there will be anything left to discuss. 👍

Author: @bwplotka aahhh, that's good to know. I'll learn more about it and update the proposal accordingly. Thanks for the tip.
